### ISSUE TITLE
Error opening consensus database: cannot allocate memory

### ISSUE THREAD

**Issue by Granitoid** [NebulousLabs/Sia#1037]
```
Mar 24 11:34:27 siad[2028]: Loading...
Mar 24 11:34:27 siad[2028]: (1/7) Loading gateway...
Mar 24 11:34:28 siad[2028]: (2/7) Loading consensus...
Mar 24 11:34:28 siad[2028]: error opening consensus database: cannot allocate memory
Mar 24 11:34:28 systemd[1]: sia.service: main process exited, code=exited, status=1/FAILURE
Mar 24 11:34:28 systemd[1]: Unit sia.service entered failed state.

$ free -h
total used free shared buffers cached
Mem: 1,7G 184M 1,5G 9M 11M 97M
-/+ buffers/cache: 76M 1,7G
Swap: 3,7G 0B 3,7G

$ ls -lA /home/sia/consensus
-rw------- 1 sia sia 2147483647 Mar 24 11:34 consensus.db
```

Source:
https://github.com/NebulousLabs/Sia-UI/releases/download/v0.5.1-beta/Sia-UI-linux-ia32.tar.gz

OS:
Linux version 3.16.0-4-amd64 (gcc version 4.8.4 (Debian 4.8.4-1) ) #1 SMP Debian 3.16.7-ckt20-1+deb8u1 (2015-12-14)

**Comment by DavidVorick** [comment]
how much disk space do you have? Unless other people on linux-32 are having similar problems, I'm guessing that this is more of a problem with your machine.

**Comment by Granitoid** [comment]
Home foler /home/sia

```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda1       225G  3.9G  210G   2% /
udev             10M     0   10M   0% /dev
tmpfs           354M  4.9M  349M   2% /run
tmpfs           884M     0  884M   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           884M     0  884M   0% /sys/fs/cgroup
/dev/sda7       232M  2.1M  214M   1% /tmp
/dev/sda6       232M   31M  185M  15% /boot
```

```
ls -lR /home/sia
/home/sia:
total 40
drwx------ 2 sia  sia  4096 Mar 23 15:46 consensus
drwx------ 2 sia  sia  4096 Mar 28 21:47 gateway
drwx------ 2 sia  sia  4096 Mar 24 03:25 host
drwx------ 2 sia  sia  4096 Mar 24 03:25 miner
drwx------ 2 sia  sia  4096 Mar 23 15:46 renter
drwx------ 2 sia  sia  4096 Mar 23 19:32 wallet

/home/sia/consensus:
total 1048512
-rw------- 1 sia sia 2147483647 Mar 28 21:39 consensus.db

/home/sia/gateway:
total 4604
-rw-r----- 1 sia sia 4701198 Mar 28 21:39 gateway.log
-rw-r--r-- 1 sia sia    7164 Mar 28 21:47 nodes.json

/home/sia/host:
total 8
-rw-r----- 1 sia sia 1214 Mar 23 19:18 host.log
-rw-r--r-- 1 sia sia 1394 Mar 24 03:25 settings.json

/home/sia/miner:
total 12
-rw-r--r-- 1 sia sia 6658 Mar 24 03:25 miner.json
-rw-r----- 1 sia sia  340 Mar 23 19:05 miner.log

/home/sia/renter:
total 8
-rw-r----- 1 sia sia 390 Mar 23 19:05 hostdb.log
-rw-r----- 1 sia sia 395 Mar 23 19:05 renter.log

/home/sia/wallet:
total 12
-rw-r--r-- 1 sia sia 452 Mar 23 15:50 Sia Wallet Encrypted Backup Seed - xxxxxxxxxxxxxxxxxx.seed
-rw-r--r-- 1 sia sia 925 Mar 23 19:12 wallet.json
-rw-r----- 1 sia sia 594 Mar 23 19:11 wallet.log
```

```
$ sudo -u sia /usr/local/bin/siad -d /home/sia -H ":9982"
Loading...
(1/7) Loading gateway...
(2/7) Loading consensus...
error opening consensus database: cannot allocate memory
```

**Comment by DavidVorick** [comment]
It seems like our database backend only supports databases of up to 2GB on 32bit machines, at this point the Sia database is starting to grow larger than 2GB, and while we could probably compress it some, it would only stay below 2GB for a while, especially as we are intending to continue growing (which means bigger blocks, which means a faster growing database).

We may have to drop support for 32bit machines in the near future. I've made an issue on our database's issue tracker to see if the maintainers are willing to support >2GB databases for 32bit machines, but I'm guessing that they are not going to be able to do so easily.

**Comment by PetrovskYYY** [comment]
Got the same issue on Ubuntu 14.04.4 LTS (32-bit)
Looks like it is the same problem:
```
$ ./siad
Loading...
(1/7) Loading gateway...
(2/7) Loading consensus...
error opening consensus database: cannot allocate memory
```
Consensus DB has grown to the same size
```
$ ls -l consensus
-rw------- 1 sysadmin sysadmin 2147483647 Апр 26 22:45 consensus.db
```
Here is output from siad upon crash (last part, because it was cut ty screen output buffer):
```
github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).AcceptBlock(0x18b61080, 0x0, 0x9e910000, 0x4e2d11ee, 0xda69e257, 0x947441d2, 0x1d3ba6e7, 0x23b919e7, 0xb4e8c63d, 0xa94bd89, ...)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/accept.go:131 +0x53
github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).threadedReceiveBlocks(0x18b61080, 0xb75dd498, 0x18c720e0, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/synchronize.go:94 +0x3a8
github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).(github.com/NebulousLabs/Sia/modules/consensus.threadedReceiveBlocks)-fm(0xb75dd498, 0x18c720e0, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/consensusset.go:117 +0x3c
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).RPC(0x18bc96c0, 0x1bd87e80, 0x12, 0x8595140, 0xa, 0x18c720b8, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/rpc.go:53 +0x3d5
created by github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).relayBlock
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/synchronize.go:225 +0x21c

goroutine 382327 [semacquire]:
sync.runtime_Syncsemacquire(0x1f2db808)
        /usr/local/go/src/runtime/sema.go:237 +0x1ef
sync.(*Cond).Wait(0x1f2db800)
        /usr/local/go/src/sync/cond.go:62 +0x85
github.com/inconshreveable/muxado/proto/buffer.(*Inbound).Read(0x2170bac0, 0x28085ff8, 0xa885c, 0xa885c, 0x0, 0x0, 0x0)
        /home/luke/go/src/github.com/inconshreveable/muxado/proto/buffer/inbound.go:155 +0x44c
github.com/inconshreveable/muxado/proto.(*Stream).Read(0x25c08a50, 0x28085ff8, 0xa885c, 0xa885c, 0xc001, 0x0, 0x0)
        /home/luke/go/src/github.com/inconshreveable/muxado/proto/stream.go:73 +0x4c
github.com/inconshreveable/muxado.(*streamAdaptor).Read(0x18c720a8, 0x28085ff8, 0xa885c, 0xa885c, 0xc001, 0x0, 0x0)
        <autogenerated>:7 +0x66
github.com/NebulousLabs/Sia/modules/gateway.(*peerConn).Read(0x18c720b0, 0x28085ff8, 0xa885c, 0xa885c, 0xc001, 0x0, 0x0)
        <autogenerated>:4 +0x68
io.ReadAtLeast(0xb75dd518, 0x18c720b0, 0x27fd6000, 0x158854, 0x158854, 0x158854, 0xafff8, 0x0, 0x0)
        /usr/local/go/src/io/io.go:298 +0xab
io.ReadFull(0xb75dd518, 0x18c720b0, 0x27fd6000, 0x158854, 0x158854, 0x27fd6000, 0x0, 0x0)
        /usr/local/go/src/io/io.go:316 +0x51
github.com/NebulousLabs/Sia/encoding.ReadPrefix(0xb75dd518, 0x18c720b0, 0x1312d00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/encoding/prefix.go:22 +0x520
github.com/NebulousLabs/Sia/encoding.ReadObject(0xb75dd518, 0x18c720b0, 0x8443220, 0x2389cb30, 0x1312d00, 0x0, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/encoding/prefix.go:28 +0x47
github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).threadedReceiveBlocks(0x18b61080, 0xb75dd498, 0x18c720b0, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/synchronize.go:85 +0x241
github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).(github.com/NebulousLabs/Sia/modules/consensus.threadedReceiveBlocks)-fm(0xb75dd498, 0x18c720b0, 0x0, 0x0)

 /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/consensusset.go:117 +0x3c
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).RPC(0x18bc96c0, 0x1f2db720, 0x12, 0x8595140, 0xa, 0x18c72070, 0x0, 0x0)
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/rpc.go:53 +0x3d5
created by github.com/NebulousLabs/Sia/modules/consensus.(*ConsensusSet).relayBlock
        /home/luke/go/src/github.com/NebulousLabs/Sia/modules/consensus/synchronize.go:225 +0x21c
```
Maybe it will help somehow.
Could this problem be solved for 32-bit mashines? if not - it will not allow 32-bit hosts to partisipate in network. and i think it is rather large number of potential hosts. Maybe it is possible to divide this database to several files, or just change file pointer type in consensus DB to 64-bit (I have several 32 bit programs on the same system that CAN use files with size more than 2Gb.

**Comment by DavidVorick** [comment]
So, one possible solution available to us is to split the database into several files. But that would take a lot of time, and isn't a priority right now. For the near future, we're going to have to drop support for 32bit machines. If it ends up being a huge blocking factor for many users, we'll figure out some way to split the database so that 32bit users can continue using Sia.

But, as far as I'm aware, aside from <$50 machines, 32bit processors have not been sold in many years. Even the raspi3 is 64bit. So I don't think it's worth a ton of effort at this stage to preserve compatibility with 32bit machines.

We may change our minds in the near future, if there is a lot of negative response to this decision. But even if we did choose to split the database, it would be several months before the code to do so could be completed.

Closing this issue for now, I wish there was a simpler way that we could support our 32bit brethren.
